### DISCUSSION TITLE
Why is Constraint\Valid not allowed to be inside of Constraint\Composite (Constraint\When)

### DISCUSSION THREAD

**Discussion by pkly** [symfony/symfony#58839] [Q&A]
Hello, 

We wanted to use `Constraint\When` along with `Constraint\Valid`, but that's not allowed. Why is that?
Our input would go into a dto like so:

```php
// load from something
public Object|null $property = null;

#[When(expression: 'property is null', constraints: [new Valid()])]
public Fallback|null $fallback = null;
```

But Constraint\Composite has a specific check just for `Valid`. Is there a proper way to validate something in a situation like this, without just re-doing the work of `Valid`?

**Comment by MatTheCat** [comment]
Hello, I guess the issue with `Valid` in composite constraints is that it would trigger an infinite loop because the `ValidValidator` would validate `$fallback`, which require the `ValidValidator`, which would validate `$fallback`…

But maybe you don’t need `When`: can `$fallback` really be invalid if `$property` is `null`?

**Reply by pkly** [reply]
In our case it is required, as the `$fallback` object can exist theoretically, and I sure know not to trust an API call. It's not used if `$property` is not `null`.

Additionally we can achieve this effect just fine with a custom assert, so that's what we'll do for now I suppose. I was just surprised, since I was working on a code review for a coworker and realized that something like `When` _probably_ exists already in symfony (we used a completely custom assert at that time).

I did consider an infinite loop, but it doesn't seem like it should happen, at least in our case. It could be possible however if you needed to check something with `When` only when a property of `$fallback` is already validated (weird scenario).
